src/dfuzzer.c

```diff
@@ -924,122 +924,54 @@ int df_get_pid(const GDBusConnection *dcon)
-                        fprintf(stderr, "%s%s[PACKAGE: %s\n", ansi_cr(), ansi_cyan(), ansi_normal());
+                if (!strstr(name, "python") && !strstr(name, "perl")) {
+                        fprintf(stderr, "%s%s[PROCESS: %s]%s\n",
+                                ansi_cr(), ansi_cyan(), name, ansi_normal());
                         return;
-                } else {    // removes cmdline arguments
-                        c = name;
-                        while (*c != ' ' && *c != '\0')
-                                c++;
-                        *c = '\0';
                 }
         }
 
-
-        fd = open("/dev/null", O_RDWR, S_IRUSR | S_IWUSR);
-        if (fd == -1) {
-                fprintf(stderr, "%s%s[PACKAGE: %s\n", ansi_cr(), ansi_cyan(), ansi_normal());
+        // if readlink failed or executable was interpret (and our target is
+        // interpreted script), try to read cmdline
+        sprintf(proc_path, "/proc/%d/cmdline", pid);
+        fd = open(proc_path, O_RDONLY);
+        if (fd <= 0) {
+                perror("open");
                 return;
         }
 
-        // backup std descriptors
-        stdoutcpy = dup(1);
-        stderrcpy = dup(2);
+        for (int i = 0;; i++) {
+                if (i >= PATH_MAX) {
+                        df_verbose("The process name was truncated\n");
+                        name[PATH_MAX] = '\0';
+                        break;
+                }
 
-        // make stdout and stderr go to fd
-        if (dup2(fd, 1) == -1) {
-                fprintf(stderr, "%s%s[PACKAGE: %s\n", ansi_cr(), ansi_cyan(), ansi_normal());
-                return;
-        }
-        if (dup2(fd, 2) == -1) {
-                fprintf(stderr, "%s%s[PACKAGE: %s\n", ansi_cr(), ansi_cyan(), ansi_normal());
-                dup2(stdoutcpy, 1);
-                close(stdoutcpy);
-                return;
-        }
-        close(fd);      // fd no longer needed
-
-
-        // Determines which package manager should be used
-        if (WEXITSTATUS(system("which rpm")) == 0)
-                sprintf(buf, "rpm -qf %s", name);
-        else if (WEXITSTATUS(system("which dpkg")) == 0)
-                sprintf(buf, "aptitude versions $(dpkg -S %s "
-                             "| sed 's/:.*//') -F %%p %%V | sed 's/Package //' "
-                             "| sed ':a;$!N;s/:\\n/-/;ta' | tr -d ' ' | sed '/^$/d' "
-                             "| tr '\n' ' ' | sed 's/$/\\n$/'" , name);
-        else {  // only rpm/dpkg are supported
-                fprintf(stderr, "%s%s[PACKAGE: %s\n", ansi_cr(), ansi_cyan(), ansi_normal());
-                // restore std descriptors
-                dup2(stdoutcpy, 1);
-                close(stdoutcpy);
-                dup2(stderrcpy, 2);
-                close(stderrcpy);
-                return;
-        }
+                ret = read(fd, (name + i), 1);
+                if (ret <= 0) {
```

Comment:
I think it should be '<' instead of '<=`. I wonder if this for loop can be replaced with something like
```
ret = read(fd, name, PATH_MAX);
if (ret < 0)
    return;
name[ret] = '\0';
```
it doesn't handle signals and kind of relies on the fact that normally there are zero bytes separating commands and their arguments but I think it's good enough considering that `dfuzzer` isn't supposed to be a mission-critical binary that can be sent to Mars or anything like that.

FWIW the PR looks good to me in its current form so if you think it's good enough feel free to press the merge button. Thanks!

Reply:
> I think it should be '<' instead of '<=`.

Indeed, thanks!

> I wonder if this for loop can be replaced with something like
> 
> ```
> ret = read(fd, name, PATH_MAX);
> if (ret < 0)
>     return;
> name[ret] = '\0';
> ```
> 
> it doesn't handle signals and kind of relies on the fact that normally there are zero bytes separating commands and their arguments but I think it's good enough considering that `dfuzzer` isn't supposed to be a mission-critical binary that can be sent to Mars or anything like that.
> 

Hehe, yeah, I also wondered if it couldn't be written in a simpler way. I guess I'll keep it there for now and focus on the rest of the code base.

Reply:
> I guess I'll keep it there for now and focus on the rest of the code base.

I think it can be trimmed significantly. For example with https://github.com/systemd/systemd-dfuzzer/commit/798ac0477d1bda1995c386baf8ead2c2d2c311a4 merged I think https://github.com/systemd/systemd-dfuzzer/blob/main/src/remove_escapes.sh can safely be removed. I haven't looked at the C code closely though. The fd and memory leaks vaglrind reports kind of make me avoid it as long as I can but I'll get there I think.